### PR TITLE
power armor

### PR DESCRIPTION
**Pull request by jjpark-kb**
## About The Pull Request
adds power armor

power armor can be upgraded by certain items

all items available in research and mech fab.

HOW TO BUILD:

- Research the necessary research nodes for the power armor and subsequent upgrades.
- Go to the mech fabricator in robotics and print all possible items under the power armor category
- Take the construct parts and place them on the skeleton
- Once a construct part is placed, tools are required to continue. Examine the skeleton to see what is required
- Once all parts are put on and all tool actions have been done, you will need a forged plate to finish the power armor
- You will now have power armor

HOW TO UPGRADE:

- Print the upgrades from the mech fabricator if you have researched them.
- Certain upgrades cost points; examine the power armor to see how many points you have left (default is 20)
- If you hit the capacity and don't want that upgrade, use a screwdriver to remove the upgrades and reset the power armor
- If you are running out of upgrade capacity, you can one-time use an anomaly to increase the capacity by 10

GYAZO GIF LINKS FOR POWERARMOR BEING WORN
https://gyazo.com/6acee7e14775d25319fc968f9f0898ed
https://gyazo.com/0c42bedef5d53b7f1babdbe868ff7e4b

## How This Contributes To The Skyrat Roleplay Experience
giving roboticists more to do rather than just mechs and borgs would be nice, especially for crew who want power armor. Additionally, it ties in content from reagent forging (requiring completed forged plates)

## Changelog

:cl:
add: added power armor
add: added power armor upgrades
/:cl:

### CODE REVIEW

**Comment by ReynTheLord** [comment]
Looks a bit unbalanced Imo,but i'm not really the best judge of these things.

**Comment by thestubborn** [comment]
haha yes

**Comment by YakumoChen** [comment]
I see crippling issues with this:
-This is going to be an immediate policy concern to prevent sec from ordering these every round. It looks all but guaranteed to be locked to code red, putting it in the same niche as mechs. Meaning they will probably never be legitimately made since it looks like these take painfully long to make.
-Given these are fully modular, I forsee there will be only one or two "meta builds" to maximize useful defenses and ignoring the rest. Something like a melee and bullet immune power armor is going to counter 90% of threats on skyrat and nothing else will ever be made.

**Comment by Gandalf2k15** [comment]
Also I refuse to allow this to be merged without digisprites

**Review by Fikou — CHANGES_REQUESTED** [review]
this should at least use associated lists instead of list[1] so the code is readable

File: modular_skyrat/modules/powerarmor/code/powerarmor.dm [hunk 1]
```diff
@@ -0,0 +1,839 @@
...
+		clothing_flags = STOPSPRESSUREDAMAGE | THICKMATERIAL
+		max_heat_protection_temperature = SPACE_HELM_MAX_TEMP_PROTECT
+		min_cold_protection_temperature = SPACE_HELM_MIN_TEMP_PROTECT
+	flash_protect = FLASH_PROTECTION_NONE
+	if(power_suit.misc_upgraded[3])
+		flash_protect = FLASH_PROTECTION_WELDER
+
+/obj/item/powerarmor_upgrade
+	name = "power armor upgrade"
+	desc = "A small item that will upgrade the power suit"
```
Comment (Fikou) [inline modular_skyrat/modules/powerarmor/code/powerarmor.dm]: grammar

File: modular_skyrat/modules/powerarmor/code/powerarmor.dm [hunk 2]
```diff
@@ -0,0 +1,839 @@
...
+		min_cold_protection_temperature = SPACE_SUIT_MIN_TEMP_PROTECT
+		if(hood && istype(hood, /obj/item/clothing/head/hooded/powerarmor))
+			var/obj/item/clothing/head/hooded/powerarmor/power_hood = hood
+			power_hood.update_upgrades()
+	light_range = 0
+	if(misc_upgraded[2])
+		light_range = 3
+		light_color = LIGHT_COLOR_LIGHT_CYAN
+
+/obj/item/clothing/suit/hooded/powerarmor/attackby(obj/item/W, mob/user, params)
```
Comment (Fikou) [inline modular_skyrat/modules/powerarmor/code/powerarmor.dm:373]: descriptive var names

File: modular_skyrat/modules/powerarmor/code/powerarmor.dm [hunk 3]
```diff
@@ -0,0 +1,839 @@
...
+	light_power = 0.75
+
+	mutant_variants = NONE
+	flags_inv = HIDEGLOVES|HIDESHOES|HIDEJUMPSUIT|HIDESEXTOY
+
+	///the maximum amount of upgrades it can have (where some upgrades can cost multiple)
+	var/upgradelimit = 20
+	///whether the upgradelimit has been up'd, by some item
+	var/upgradeboosted = FALSE
+	//the cooldown for
```
Comment (Fikou) [inline modular_skyrat/modules/powerarmor/code/powerarmor.dm]: for

Comment (Fikou) [reply modular_skyrat/modules/powerarmor/code/powerarmor.dm]: also all of these vars should get autodoc comments

File: modular_skyrat/modules/powerarmor/code/powerarmor.dm [hunk 4]
```diff
@@ -0,0 +1,839 @@
...
+		living_wearer.Stun(5 SECONDS)
+		living_wearer.adjustFireLoss(25)
+		to_chat(living_wearer, span_warning("[src] short-circuits, hurting you in the process!"))
+
+/obj/item/clothing/suit/hooded/powerarmor/AltClick(mob/user)
+	. = ..()
+	var/list/selection = list()
+	if(misc_upgraded[5])
+		selection += "storage"
+	var/get_choice = tgui_input_list(user, "Choose which option to selection", "Selection Menu", selection)
```
Comment (Fikou) [inline modular_skyrat/modules/powerarmor/code/powerarmor.dm:289]: but theres only 1 choice youll ever get, storage

**Comment by Fikou** [comment]
also there is a LOT of magic numbers here that should be turned into defines (and the code in general is not very sleek or anything but i guess its fine??)

File: modular_skyrat/modules/powerarmor/code/powerarmor.dm (hunk 4 above)
Comment (jjpark-kb) [reply modular_skyrat/modules/powerarmor/code/powerarmor.dm:289]: open to allow expansion.

File: modular_skyrat/modules/powerarmor/code/powerarmor.dm (hunk 1 above)
Comment (jjpark-kb) [reply modular_skyrat/modules/powerarmor/code/powerarmor.dm]: whats the issue?

Comment (Fikou) [reply modular_skyrat/modules/powerarmor/code/powerarmor.dm]: it kinda sounds weird in general but at least there should be a period at the end

Comment (jjpark-kb) [reply modular_skyrat/modules/powerarmor/code/powerarmor.dm]: ah

**Comment by GoldenAlpharex** [comment]
Okay, Jake, I love the idea, but please. One way or the other, you need to be detailed on how it works, either on the wiki directly, or in the PR body, so someone can take it to the wiki, so it's not a nightmare for mentors to figure out how this works to answer questions. It's probably very nice, but 95% of the people won't have a clue how this works, because this is one of the most vague PR body I've seen in a while from you, lol

**Comment by jjpark-kb** [comment]
HOLD THE PHONE

**Comment by YakumoChen** [comment]
why was this merged already

**Comment by chefferz** [comment]
After watching a round this is absolutely 100% ridiculous.
You go a ridiculous speed. Easily clearing 20-30 tiles per second.
This is under a basic unlock. Why? At the minimum this should be extremely expensive and exponentially difficult to obtain if it remains this overpowered.

**Comment by chefferz** [comment]
After watching a round.
You go a ridiculous speed. Easily clearing 20-30 tiles per second.
This is under a basic unlock. Why? At the minimum this should be extremely expensive and exponentially difficult to obtain if it remains this overpowered.
If you kept the speed in very very short bursts that drained a cell that would make sense, but this is giving you literal meth+ running speed with zero downsides.

**Pull request merged**